src/main/java/com/google/api/gax/grpc/RetryingCallable.java

```diff
@@ -29,226 +29,151 @@
+      return oldCtx;
     }
+    return newCtx;
   }
 
-  private class RetryingResultFuture extends AbstractApiFuture<ResponseT> {
-    private volatile Future<?> activeFuture = null;
-    private final Object syncObject = new Object();
+  private static class GrpcRetryCallable<RequestT, ResponseT> implements Callable<ResponseT> {
+    private final FutureCallable<RequestT, ResponseT> callable;
+    private final RequestT request;
 
-    @Override
-    protected void interruptTask() {
-      synchronized (syncObject) {
-        activeFuture.cancel(true);
-      }
-    }
+    private volatile RetryingFuture<ResponseT> externalFuture;
+    private volatile CallContext callContext;
 
-    @Override
-    public boolean set(@Nullable ResponseT value) {
-      synchronized (syncObject) {
-        return super.set(value);
-      }
+    private GrpcRetryCallable(
+        FutureCallable<RequestT, ResponseT> callable, RequestT request, CallContext callContext) {
+      this.callable = callable;
+      this.request = request;
+      this.callContext = callContext;
     }
 
-    @Override
-    public boolean setException(Throwable throwable) {
-      synchronized (syncObject) {
-        if (throwable instanceof CancellationException) {
-          super.cancel(false);
-          return true;
-        } else {
-          return super.setException(throwable);
-        }
-      }
+    private void setExternalFuture(RetryingFuture<ResponseT> externalFuture) {
+      this.externalFuture = externalFuture;
     }
 
-    private void scheduleNext(
-        UnaryCallable.Scheduler executor, Runnable retryer, long delay, TimeUnit unit) {
-      synchronized (syncObject) {
-        if (!isCancelled()) {
-          activeFuture = executor.schedule(retryer, delay, TimeUnit.MILLISECONDS);
+    @Override
+    public ResponseT call() {
+      callContext =
+          getCallContextWithDeadlineAfter(
+              callContext, externalFuture.getAttemptSettings().getRpcTimeout());
+
+      try {
+        externalFuture.setAttemptFuture(new NonCancelableFuture<ResponseT>());
+        if (externalFuture.isDone()) {
+          return null;
         }
+        ApiFuture<ResponseT> internalFuture = callable.futureCall(request, callContext);
+        externalFuture.setAttemptFuture(new ApiFutureToListenableFuture<>(internalFuture));
+      } catch (Throwable e) {
+        externalFuture.setAttemptFuture(Futures.<ResponseT>immediateFailedFuture(e));
+        throw e;
       }
-    }
 
-    public void issueCall(
-        RequestT request,
-        CallContext deadlineContext,
-        RetryingCallable<RequestT, ResponseT>.Retryer retryer) {
-      synchronized (syncObject) {
-        if (!isCancelled()) {
-          ApiFuture<ResponseT> callFuture = callable.futureCall(request, deadlineContext);
-          ApiFutures.addCallback(callFuture, retryer);
-          activeFuture = callFuture;
-        }
-      }
+      return null;
```

Comment:
Yes it does feel wrong. It is essentially a hack, but when you have to make compatible two incompatible things, things like this happen. I still believe it is a very mild hack, so we can leave with it. 

The reason why this exist is same as described in the previous comment - the substitution of the future wrapped by RetryingFuture in the middle of the process - swapping ScheduledExecutorFutureTask with another asynchronous ApiFuture. 

In other words, what happens is chaining of **two asynchronous** calls, and since the second call is also asynchronous, the first one cannot return anything, because it does not know the actual operation result yet, and somebody has to wait for it to happen. The one who is waiting is the outside code, not the current intermediate task (otherwise it would consume threadpool thread).